ingestion_server/ingestion_server/cleanup.py

```diff
@@ -266,30 +267,89 @@ def _clean_data_worker(rows, temp_table, sources_config, all_fields: list[str]):
     worker_conn.commit()
     write_cur.close()
     worker_conn.close()
-    end_time = time.time()
+    end_time = time.perf_counter()
     total_time = end_time - start_time
     log.info(f"Worker finished batch in {total_time}")
     return cleaned_values
 
 
-def save_cleaned_data(result: dict) -> dict[str, int]:
-    log.info("Saving cleaned data...")
-    start_time = time.time()
+@dataclass
+class FieldBuffered:
+    part: int
+    rows: list[tuple[str, str]]
 
-    cleanup_counts = {field: len(items) for field, items in result.items()}
-    for field, cleaned_items in result.items():
-        # Skip the tag field because the file is too large and fills up the disk
-        if field == "tag":
-            continue
-        if cleaned_items:
-            with open(f"{field}.tsv", "a") as f:
-                csv_writer = csv.writer(f, delimiter="\t")
-                csv_writer.writerows(cleaned_items)
 
-    end_time = time.time()
-    total_time = end_time - start_time
-    log.info(f"Finished saving cleaned data in {total_time},\n{cleanup_counts}")
-    return cleanup_counts
+class CleanDataUploader:
+    # Number of lines to keep in memory before writing to S3
+    buffer_size: int
+
+    s3_path = "shared/data-refresh-cleaned-data"
+
+    buffer = {
+        field: FieldBuffered(part=1, rows=[])
+        for field in _cleanup_config["tables"]["image"]["sources"]["*"]["fields"]
+    }
+
+    def __init__(self):
+        self.buffer_size = config("CLEANUP_BUFFER_SIZE", default=10_000_000, cast=int)
+        bucket_name = config("OPENVERSE_BUCKET", default="openverse-catalog")
+        try:
+            self.s3 = self._get_s3_resource()
+            self.s3_bucket = self.s3.Bucket(bucket_name)
+            self.date = time.strftime("%Y-%m-%d")
+        except Exception as e:
+            log.error(f"Error connecting to S3 or creating bucket: {e}")
+            self.s3 = None
+            self.s3_bucket = None
+
+    @staticmethod
+    def _get_s3_resource():
+        if config("ENVIRONMENT", default="local") == "local":
+            return boto3.resource(
+                "s3",
+                endpoint_url=config("AWS_S3_ENDPOINT", default="http://s3:5000"),
+                aws_access_key_id=config("AWS_ACCESS_KEY_ID", default="test_key"),
+                aws_secret_access_key=config(
+                    "AWS_SECRET_ACCESS_KEY", default="test_secret"
+                ),
+            )
+
+        return boto3.resource(
+            "s3", region_name=config("AWS_REGION", default="us-east-1")
+        )
+
+    def _upload_to_s3(self, field: str):
+        part_number = self.buffer[field].part
+        log.info(f"Uploading file part {part_number} of `{field}` to S3...")
+        s3_file_name = f"{self.s3_path}/{self.date}_{field}_{part_number}.tsv"
+        tsv_file = f"{field}.tsv"
+        with open(tsv_file, "w") as f:
+            csv_writer = csv.writer(f, delimiter="\t")
```

Comment:
It looks like this current set of settings will cause the `url` column (at least when I tested it) to be quoted. Since we're using tab delimiting, we shouldn't need to worry about quoting fields. It may also make ingestion into Postgres harder down the line too, if we have to try and unquote the fields before inserting them into the database after reading the file! I'd recommend the following[^1]:

```suggestion
            csv_writer = csv.writer(f, delimiter="\t", quoting=csv.QUOTE_NONE)
```

[^1]: https://docs.python.org/3/library/csv.html#csv.QUOTE_NONE